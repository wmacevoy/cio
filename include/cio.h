#pragma once

#include <stdint.h>

#ifdef __cplusplus
extern "C" {
#endif

  struct CIOStruct;
  typedef struct CIOStruct CIO;
  typedef int (*CIOReadPtr)(CIO *me);
  typedef int (*CIOWritePtr)(CIO *me,int value);
  typedef int (*CIOPeekPtr)(CIO *me,int offset);
  typedef void (*CIOClosePtr)(CIO *me);
  struct CIOStruct {
    CIOReadPtr read;
    CIOWritePtr write;
    CIOPeekPtr peek;
    CIOClosePtr close;
    int reads;
    int writes;
  };
  void CIOInit(CIO *me);

  int CIORead(void *me);

  int CIOWrite(void *me, int data);

  int CIOPeek(void *me, int offset);

  void CIOClose(void *me);

  int CIOGetReads(void *me);

  int CIOGetWrites(void *me);

  struct CIOArrayStruct;
  typedef struct CIOArrayStruct CIOArray;

  typedef int (*CIOArrayElementReadPtr)(void *location);
  typedef void (*CIOArrayElementWritePtr)(void *location, int value);

  struct CIOArrayStruct {
    CIO base;
    CIOArrayElementReadPtr elementRead;
    CIOArrayElementWritePtr elementWrite;
    int elementSize;
    void *data;
    int position;
    int size;
    int capacity;
    int maxCapacity;
  };

  void CIOArrayInit(CIOArray *me,
		    CIOArrayElementReadPtr elementRead,
		    CIOArrayElementWritePtr elementWrite,		    
		    int elementSize,
		    void *data,
		    int size,
		    int capacity,
		    int maxCapacity);

  void CIOArrayU8Init(CIOArray *me,
		    uint8_t *data,
		    int size,
		    int capacity,
		    int maxCapacity);
  
  void CIOArrayU16Init(CIOArray *me,
		    uint16_t *data,
		    int size,
		    int capacity,
		    int maxCapacity);
  
  void CIOArrayU32Init(CIOArray *me,
		    uint32_t *data,
		    int size,
		    int capacity,
		    int maxCapacity);
  

#ifdef __cplusplus
} // extern "C"
#endif  
