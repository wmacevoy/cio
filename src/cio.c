#include <stdlib.h>
#include <string.h>

#include "cio.h"

int CIORead(void *me) {
  int status = ((CIO*)me)->read((CIO*)me);
  if (status != -1) { ++(((CIO*)me)->reads); }
  return status;
}

int CIOWrite(void *me, int data) {
  int status = -1;
  if (data != -1) {
    status = ((CIO*)me)->write((CIO*)me,data);
    if (status != -1) { ++(((CIO*)me)->writes); }
  } else {
    ((CIO*)me)->close((CIO*)me);
  }
  return status;
}

int CIOPeek(void *me, int offset) {
  return ((CIO*)me)->peek((CIO*)me,offset);
}

void CIOClose(void *me) {
  ((CIO*)me)->close((CIO*)me);
}

int CIOGetReads(void *me) { ((CIO*)me)->reads; }

int CIOGetWrites(void *me) { ((CIO*)me)->writes; }  

static int CIOBaseRead(CIO *me) { return -1; }
static int CIOBaseWrite(CIO *me, int data) { return 0; }
static int CIOBasePeek(CIO *me, int offset) { return -1; }
static void CIOBaseClose(CIO *me) {}



void CIOInit(CIO *me) {
  me->read = &CIOBaseRead;
  me->write = &CIOBaseWrite;
  me->peek = &CIOBasePeek;
  me->close = &CIOBaseClose;
  me->reads = 0;
  me->writes = 0;
}

static int CIOArrayPeek(CIOArray *me, int offset) {
  int position = me->position+offset;
  if (0 <= position && position < me->size) {
    unsigned char *location = ((unsigned char*) me->data) + position*me->elementSize;
    int value = me->elementRead(location);
    return value;
  } else {
    return -1;
  }
}

static int CIOArrayRead(CIOArray *me) {
  int value = CIOArrayPeek(me,0);
  if (value != -1) {
    ++(me->position);
  }
  return value;
}

static int CIOArrayWrite(CIOArray *me, int value) {
  int position = me->position;
  if (position >= me->capacity) {
    if (position < me->maxCapacity) {
      int newCapacity = 3*position/2;
      if (newCapacity < 32) newCapacity = 32;
      if (newCapacity > me->maxCapacity) newCapacity=me->maxCapacity;
      void *newData = realloc(me->data,me->elementSize*newCapacity);
      if (newData != NULL) {
	memset(((unsigned char*)newData)+me->elementSize*me->capacity,
	       0,
	       me->elementSize*(newCapacity-me->capacity));
	me->data = newData;
	me->capacity = newCapacity;
      } else {
	return -1;
      }
    }
  }
  if (me->size <= me->position) {
    me->size = me->position+1;
  }
  void *location = ((unsigned char*) me->data)+me->position*me->elementSize;
  me->elementWrite(location,value);
  ++me->position;
  return 0;
}


static void CIOArrayClose(CIOArray *me) {
}

void CIOArrayInit(CIOArray *me,
		  CIOArrayElementReadPtr elementRead,
		  CIOArrayElementWritePtr elementWrite,
		  int elementSize,
		  void *data,
		  int size,
		  int capacity,
		  int maxCapacity) {
  CIOInit(&me->base);
  me->base.read=(CIOReadPtr)&CIOArrayRead;
  me->base.write=(CIOWritePtr)&CIOArrayWrite;
  me->base.peek=(CIOPeekPtr)&CIOArrayPeek;
  me->base.close=(CIOClosePtr)&CIOArrayClose;
  me->elementRead = elementRead;
  me->elementWrite = elementWrite;
  me->elementSize = elementSize;
  me->data=data;
  me->position=0;
  me->size=size;
  me->capacity=capacity;
  me->maxCapacity=me->maxCapacity;
}

int CIOArrayElementReadU8(void *location) {
  return *(uint8_t*)location;
}

void CIOArrayElementWriteU8(void *location, int value) {
  *(uint8_t*)location = value;
}

int CIOArrayElementReadU16(void *location) {
  return *(uint16_t*)location;
}

void CIOArrayElementWriteU16(void *location, int value) {
  *(uint16_t*)location = value;
}

int CIOArrayElementReadU32(void *location) {
  return *(uint32_t*)location;
}

void CIOArrayElementWriteU32(void *location, int value) {
  *(uint32_t*)location = value;
}

void CIOArrayU8Init(CIOArray *me,
		    uint8_t *data,
		    int size,
		    int capacity,
		    int maxCapacity) {
  CIOArrayInit(me,
	       CIOArrayElementReadU8,CIOArrayElementWriteU8,sizeof(uint8_t),
	       data,size,capacity,maxCapacity);
}
  
void CIOArrayU16Init(CIOArray *me,
		     uint16_t *data,
		     int size,
		     int capacity,
		     int maxCapacity) {
  CIOArrayInit(me,
	       CIOArrayElementReadU16,CIOArrayElementWriteU16,sizeof(uint16_t),
	       data,size,capacity,maxCapacity);
  
}

void CIOArrayU32Init(CIOArray *me,
		     uint32_t *data,
		     int size,
		     int capacity,
		     int maxCapacity) {
  CIOArrayInit(me,
	       CIOArrayElementReadU32,CIOArrayElementWriteU32,sizeof(uint32_t),
	       data,size,capacity,maxCapacity);
}


