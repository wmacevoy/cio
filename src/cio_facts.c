#include <limits.h>
#include "facts.h"
#include "cio.h"
#include "utf8.h"

FACTS(CIOBase) {
  CIO io;
  CIOInit(&io);
  FACT(CIOGetReads(&io),==,0);
  FACT(CIOGetWrites(&io),==,0);  
  FACT(CIORead(&io),==,-1);
  FACT(CIOPeek(&io,0),==,-1);  
  FACT(CIOWrite(&io,0),==,0);
  FACT(CIOGetReads(&io),==,0);
  FACT(CIOGetWrites(&io),==,1);  
  FACT(CIOWrite(&io,-1),==,-1);
  FACT(CIOGetReads(&io),==,0);
  FACT(CIOGetWrites(&io),==,1);  
  CIOClose(&io);
}

FACTS(CIOArrayU8Fixed) {
  uint8_t a[5] = {99,0,0,0,99};

  CIOArray io;
  CIOArrayU8Init(&io,a+1,0,3,3,3);

  FACT(CIOWrite(&io,10),==,0);  FACT(a[1],==,10);
  FACT(CIOWrite(&io,20),==,0);  FACT(a[2],==,20);
  FACT(CIOWrite(&io,30),==,0);  FACT(a[3],==,30);
  FACT(CIOWrite(&io,40),==,-1);
  FACT(a[0],==,99); FACT(a[4],==,99);
  FACT(io.data,==,(void*)(a+1));

  CIOClose(&io);
  
}

FACTS(CIOArrayU16Dynamic) {
  uint16_t a[3];
  CIOArray io;
  int n = 1000;
  CIOArrayU16Init(&io,a,0,0,3,n);

  for (int i=0; i<=n; ++i) {
    FACT(CIOWrite(&io,10*i),==,(i < n) ? 0 : -1);
    if (i < n) {
      FACT(((uint16_t*)io.data)[i],==,10*i);
    }
  }
  
  CIOClose(&io);  
}

FACTS(CIOUTF8In) {
  char buf[1024];
  char *p=buf;
  for (int k=0; k<10; ++k) {
    for (int len=1; len<=4; ++len) {
      int c = (1<<(5*len))|(k+1);
      FACT(utf8enclen(c),==,len);
      utf8encval(p,c,len);
      p += len;
    }
  }
  *p = 0;
  CIOArray u8;
  CIOArrayConstU8Init(&u8,(uint8_t*)buf,0,p-buf);
  CIOUTF8 utf8;
  CIOUTF8Init(&utf8,&u8.base);
  for (int k=0; k<10; ++k) {
    for (int len=1; len<=4; ++len) {
      int c = (1<<(5*len))|(k+1);
      FACT(CIOPeek(&utf8,k*4+(len-1)),==,c);
    }
  }
  for (int k=0; k<10; ++k) {
    for (int len=1; len<=4; ++len) {
      int c = (1<<(5*len))|(k+1);
      FACT(CIORead(&utf8),==,c);
    }
  }

  CIOClose(&utf8);
  CIOClose(&u8);
  
}

FACTS(CIOUTF8Out) {
  char buf[1024];
  char *p=buf;
  for (int k=0; k<10; ++k) {
    for (int len=1; len<=4; ++len) {
      int c = (1<<(5*len))|(k+1);
      FACT(utf8enclen(c),==,len);
      utf8encval(p,c,len);
      p += len;
    }
  }
  *p = 0;

  CIOArray u8;
  CIOArrayU8Init(&u8,NULL,0,0,0,INT_MAX);
  CIOUTF8 utf8;
  CIOUTF8Init(&utf8,&u8.base);  

  for (int k=0; k<10; ++k) {
    for (int len=1; len<=4; ++len) {
      int c = (1<<(5*len))|(k+1);
      FACT(CIOWrite(&utf8,c),==,0);
    }
  }
  FACT(CIOWrite(&utf8,0),==,0);

  FACT(strcmp(buf,(char*)u8.data),==,0);

  CIOClose(&utf8);
  CIOClose(&u8);
}

FACTS(CIOFILEIn) {
  FILE *tmp = tmpfile();
  fprintf(tmp,"%s","abcdefghijklmnopqrstuvwxyz");
  fseek(tmp,0,SEEK_SET);
  CIOFILE ci;
  CIOFILEInit(&ci,tmp,0);
  for (int i='a'; i<='z'; ++i) {
    FACT(CIOPeek(&ci,i-'a'),==,i);
  }
  for (int i='a'; i<='z'; ++i) {
    FACT(CIORead(&ci),==,i);
  }
  CIOClose(&ci);
}

FACTS(CIOFILEOut) {
  FILE *tmp = tmpfile();
  CIOFILE co;
  CIOFILEInit(&co,tmp,0);

  for (int i='a'; i<='z'; ++i) {
    FACT(CIOWrite(&co,i),==,0);
  }
  fseek(tmp,0,SEEK_SET);
  for (int i='a'; i<='z'; ++i) {
    FACT(fgetc(tmp),==,i);
  }
  CIOClose(&co);
}

FACTS_REGISTER_ALL() {
    FACTS_REGISTER(CIOBase);
    FACTS_REGISTER(CIOArrayU8Fixed);
    FACTS_REGISTER(CIOArrayU16Dynamic);
    FACTS_REGISTER(CIOUTF8In);
    FACTS_REGISTER(CIOUTF8Out);
    FACTS_REGISTER(CIOFILEIn);
    FACTS_REGISTER(CIOFILEOut);
}

FACTS_MAIN
