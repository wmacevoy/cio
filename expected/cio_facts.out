src/cio_facts.c 6: CIOBase facts check started
src/cio_facts.c 6: CIOBase facts check ended
src/cio_facts.c 22: CIOArrayU8Fixed facts check started
src/cio_facts.c 22: CIOArrayU8Fixed facts check ended
src/cio_facts.c 39: CIOArrayU16Dynamic facts check started
src/cio_facts.c 39: CIOArrayU16Dynamic facts check ended
src/cio_facts.c 55: CIOUTF8In facts check started
src/cio_facts.c 55: CIOUTF8In facts check ended
src/cio_facts.c 89: CIOUTF8Out facts check started
src/cio_facts.c 89: CIOUTF8Out facts check ended
src/cio_facts.c 121: CIOFILEIn facts check started
src/cio_facts.c 121: CIOFILEIn facts check ended
src/cio_facts.c 136: CIOFILEOut facts check started
src/cio_facts.c 136: CIOFILEOut facts check ended
facts summary.
facts check CIOBase [1;32mpassed[0m
facts check CIOArrayU8Fixed [1;32mpassed[0m
facts check CIOArrayU16Dynamic [1;32mpassed[0m
facts check CIOUTF8In [1;32mpassed[0m
facts check CIOUTF8Out [1;32mpassed[0m
facts check CIOFILEIn [1;32mpassed[0m
facts check CIOFILEOut [1;32mpassed[0m
2327 (100.0%) truths and 0 (0.0%) fictions checked.
